File: implementations/software/simulations/js/recallTest.js
/**
 * recallTest.js — Recall Propagation Test for Ce Protocol (JavaScript)
 *
 * Purpose:
 *   Test recall propagation across nodes by seeding an observation
 *   in one node and observing how it spreads through hint exchange.
 *
 * Non-normative. Identity-free.
 */

import { setupNetwork } from "./multiNodeSim.js";

// --------------------------------------------------
// Recall propagation test
// --------------------------------------------------

function recallPropagationTest({
  nodeCount = 4,
  steps = 5,
  delayMs = 400
} = {}) {
  const nodes = setupNetwork(nodeCount);

  console.log("\n=== Recall Propagation Test ===");

  // Seed initial observation in first node
  const initialObservation = {
    type: "observation",
    sensorType: "temperature",
    value: 25.0,
    timestamp: Date.now()
  };

  nodes[0].memory.push(initialObservation);
  console.log("Initial observation seeded in Node 0:", initialObservation);

  let step = 0;

  function stepSimulation() {
    step += 1;
    console.log(`\n--- Step ${step} ---`);

    // Random local observations
    nodes.forEach(node => {
      if (Math.random() < 0.3) {
        const value = Number((20 + Math.random() * 10).toFixed(2));
        const obs = node.generateObservation("temperature", value);
        console.log("Generated observation:", obs);
      }
    });

    // Propagate recall hints
    nodes.forEach(node => node.sendHintsToPeers());

    // Display memory state
    nodes.forEach((node, index) => {
      console.log(
        `Node ${index} memory (${node.memory.length} obs):`,
        node.memory
      );
    });

    if (step < steps) {
      setTimeout(stepSimulation, delayMs);
    }
  }

  stepSimulation();
}

// --------------------------------------------------
// Entry point
// --------------------------------------------------

if (typeof window === "undefined") {
  recallPropagationTest({ nodeCount: 4, steps: 5 });
}

// For browser use, call recallPropagationTest() manually
export { recallPropagationTest };

File: implementations/software/simulations/Python/recall_test.py
"""
recall_test.py — Recall Propagation Test for Ce Protocol

Purpose:
    Test the propagation of observations (recall hints) across multiple nodes
    to verify that recall works as intended in a minimal identity-free network.
"""

import random
import time
from multi_node_sim import Node, setup_network

def test_recall_propagation(node_count=4, steps=5):
    nodes = setup_network(node_count)
    
    print("\n=== Recall Propagation Test ===")
    
    # Step 0: Initial observation only on node 0
    initial_obs = nodes[0].generate_observation("temperature", 25.0)
    print(f"Initial observation generated on Node 0: {initial_obs}")
    
    for step in range(1, steps+1):
        print(f"\n--- Simulation step {step} ---")
        
        # Randomly choose nodes to generate new observations
        for node in nodes:
            if random.random() < 0.3:  # 30% chance to generate an observation
                value = round(random.uniform(20, 30), 2)
                obs = node.generate_observation("temperature", value)
                print(f"Node generated observation: {obs}")
        
        # Propagate hints among peers
        for node in nodes:
            node.send_hint_to_peers()
        
        # Display memory of each node
        for i, node in enumerate(nodes):
            print(f"Node {i} memory ({len(node.memory)} obs):")
            for obs in node.memory:
                print(f"  {obs}")
        
        time.sleep(0.3)

if __name__ == "__main__":
    test_recall_propagation(node_count=4, steps=5)
